### DISCUSSION TITLE
Talk to /dev/hidraw dev

### DISCUSSION THREAD

**Discussion by thowabu** [jblance/mpp-solar#332] [Q&A]
Hello,

why is it necessary to split the message in eight byte groups and send with an 0,350s pause in between?
Can You explain me how do You discoverd this "special" transmission?

Do You know if the RS232 and the native USB are feed into the same port on the inverter?

**Comment by jblance** [comment]
8 bytes per send was based on an approach stolen from the internet - not sure it is valid and there are definitely some problems with the mpp-solar hidraw code (currently waiting on my new inverter build to fix)
rs232 are not the same port on any of the inverters I am aware of (they are quite different interfaces at a physical level) though the actual port for rs232 on the inverter can be different (my inverters use RJ45 as the port plug)

**Reply by Saentist** [reply]
> 8 bytes per send was based on an approach stolen from the internet - not sure it is valid and there are definitely some problems with the mpp-solar hidraw code (currently waiting on my new inverter build to fix) rs232 are not the same port on any of the inverters I am aware of (they are quite different interfaces at a physical level) though the actual port for rs232 on the inverter can be different (my inverters use RJ45 as the port plug)

Let me insert some clarifications,
Universal Serial Bus (USB) and RS232 are two different serial ports.
The fact is that the same command sent to the inverter can return a result or a NAK over both.

The RG45 port on the inverter is not a lan port, the correct name is a "Console port", 
you can find one on communication equipment, 
although in recent years there have been attempts to replace it with USB due to "clever staff".

**Reply by thowabu** [reply]
I can only speak for the FSP MPI 10k and the MPPSolar5048mkx.

On the MPI 10k the RS232 and USB Port showed to be internaly combined.
Communication on the USB Port parallel send through the RS232 to.
(Inverter TX)
I can read the Answers on the RS232 per cutecom.

The communication over RS232 isn't difficult.
But send 8 Byte Packets (in C but the Kernel Calls for read/write etc. should be the same) doesn't work very well.
I discoverd someone who used libusb to interact with the USB Port directly.
Until now I wasn't sucessfull.

I did some usbmon - wireshark dump and working on.
USB is new to me.

But :

----
Unknown Device
Speed: 1.5Mb/s (low)
Bus:   3
Address:   2
USB Version:  1.10
Device Class: 00(>ifc )
Device Subclass: 00
Device Protocol: 00
Maximum Default Endpoint Size: 8
Number of Configurations: 1
Vendor Id: 0665
Product Id: 5161
Revision Number:  0.02

Config Number: 1
	Number of Interfaces: 1
	Attributes: a0
	MaxPower Needed: 100mA

	Interface Number: 0
		Name: (none)
		Alternate Number: 0
		Class: 03(HID  ) 
		Sub Class: 00
		Protocol: 00
		Number of Endpoints: 1

			Endpoint Address: 81
			Direction: in
			Attribute: 3
			Type: Int.
			Max Packet Size: 8
			Interval: 12ms
---

Here You can see what communication (Packetsize) the Cypress Chip is capable...

**Reply by Saentist** [reply]
@thowabu try use `update-pciids`

``` shell
$ lsusb -d 0665:5161 -v

Bus 004 Device 004: ID 0665:5161 Cypress Semiconductor USB to Serial
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0         8
  idVendor           0x0665 Cypress Semiconductor
  idProduct          0x5161 USB to Serial
  bcdDevice            0.02
  iManufacturer           3 STMicroelectronics
  iProduct                1 HID in FS Mode
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      27
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              32
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0001
  Self Powered

12: USB 00.0: 0000 Unclassified device
  [Created at usb.122]
  Unique ID: PYMB.RRD0lKfXr30
  Parent ID: zPk0.RTX9xWW_uz4
  SysFS ID: /devices/pci0000:00/0000:00:1d.2/usb4/4-1/4-1:1.0
  SysFS BusID: 4-1:1.0
  Hardware Class: unknown
  Model: "Cypress USB to Serial"
  Hotplug: USB
  Vendor: usb 0x0665 "Cypress Semiconductor"
  Device: usb 0x5161 "USB to Serial"
  Revision: "0.02"
  Driver: "usbhid"
  Driver Modules: "usbhid"
  Speed: 12 Mbps
  Module Alias: "usb:v0665p5161d0002dc00dsc00dp00ic03isc00ip00in00"
  Driver Info #0:
    Driver Status: usbhid is active
    Driver Activation Cmd: "modprobe usbhid"
  Config Status: cfg=new, avail=yes, need=no, active=unknown
  Attached to: #8 (Hub)

```
